File: solutions/clojure/protein-translation/1/src/protein_translation.clj
(ns protein-translation)


(def codons
  {"Methionine" ["AUG"]
   "Phenylalanine" ["UUU" "UUC"]
   "Leucine" ["UUA" "UUG"]
   "Serine" ["UCU" "UCC" "UCA" "UCG"]
   "Tyrosine" ["UAU" "UAC"]
   "Cysteine" ["UGU" "UGC"]
   "Tryptophan" ["UGG"]
   "STOP" ["UAA" "UAG" "UGA"]})


(def ->protein
  (->> codons
       (mapcat (fn [[p cs]] (map vector cs (repeat p))))
       (into {})))


(def translate-codon ->protein)


(defn translate-rna [s]
  (->> s
       (partition 3)
       (map (partial apply str))
       (map ->protein)
       (take-while (partial not= "STOP"))))

File: solutions/csharp/protein-translation/1/ProteinTranslation.cs
public static class ProteinTranslation
{
    private const string Stop = "STOP";

    private static readonly Dictionary<string, string> ProteinsByCodon = new()
    {
        ["AUG"] = "Methionine",
        ["UUU"] = "Phenylalanine",
        ["UUC"] = "Phenylalanine",
        ["UUA"] = "Leucine",
        ["UUG"] = "Leucine",
        ["UCU"] = "Serine",
        ["UCC"] = "Serine",
        ["UCA"] = "Serine",
        ["UCG"] = "Serine",
        ["UAU"] = "Tyrosine",
        ["UAC"] = "Tyrosine",
        ["UGU"] = "Cysteine",
        ["UGC"] = "Cysteine",
        ["UGG"] = "Tryptophan",
        ["UAA"] = Stop,
        ["UAG"] = Stop,
        ["UGA"] = Stop
    };

    public static string[] Proteins(string strand) =>
        strand.Chunk(3)
            .Select(ProteinFor)
            .TakeWhile(protein => protein != Stop)
            .ToArray();

    private static string ProteinFor(char[] chunk) =>
        ProteinsByCodon[new string(chunk)];
}
